### PR TITLE
Feature: Integrations Setup

### PR DESCRIPTION
**Pull request by alashchev17**
# Feature: Integrations Setup

## Description

- Provides the UI for configuration of integrations and command-line services & tools.
- Contains setup of a library for verbose debugging (npm package debug)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactoring (no functional changes, only code improvements)
- [ ] Documentation update

### Main integrations setup screen
Main integrations setup screen
### PostgreSQL setup screen
PostgreSQL setup screen
### Docker Containers section related to integration
Docker Containers section related to integration

## Checklist

- [x] My code follows the code style of this project.
- [x] I have performed a self-review of my code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I have updated the documentation where necessary.

### CODE REVIEW

File: src/components/IntegrationsView/IntegrationsView.tsx [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+    return (
+      integrationsMap.integrations.find(
+        (integration) =>
+          integration.project_path === currentThreadIntegration.projectPath &&
+          integration.integr_name === currentThreadIntegration.integrationName,
+      ) ?? null
+    );
+  }, [currentThreadIntegration, integrationsMap]);
+
+  // TBD: what if they went home then came back to integrations?
```
Comment (MarcMcIntosh) [inline src/components/IntegrationsView/IntegrationsView.tsx:80]: good point.

File: src/components/IntegrationsView/JSONFormStyles.css [hunk 2]
```diff
@@ -0,0 +1,29 @@
+fieldset {
```
Comment (MarcMcIntosh) [inline src/components/IntegrationsView/JSONFormStyles.css:1]: is this file used anywhere?

File: src/features/Chat/Thread/reducer.ts [hunk 3]
```diff
@@ -182,4 +191,28 @@ export const chatReducer = createReducer(initialState, (builder) => {
     state.thread.title = action.payload.title;
     state.thread.isTitleGenerated = action.payload.isTitleGenerated;
   });
+
+  builder.addCase(newIntegrationChat, (state, action) => {
+    // TODO: find out about tool use
+    const next = createInitialState("agent", action.payload.integration);
+    next.thread.integration = action.payload.integration;
+    next.thread.messages = action.payload.messages;
+
```
Comment (MarcMcIntosh) [inline src/features/Chat/Thread/reducer.ts:200]: Delete these comments

File: src/services/refact/docker.ts [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+
+  if (typeof container.image !== "string") return false;
+  if (typeof container.working_dir !== "string") return false;
+  if (!isDockerLabels(container.labels)) return false;
+  if (!isDockerPorts(container.ports)) return false;
+  return true;
+}
+
+function isDockerLabels(json: unknown): json is DockerLabels {
+  // Since DockerPorts is defined as NonNullable<unknown>, we don't have specific structure to validate. Just checking, that it's not null | undefined
```
Comment (MarcMcIntosh) [inline src/services/refact/docker.ts:294]: Are these unknown or likely to change?

File: src/services/refact/docker.ts [hunk 5]
```diff
@@ -0,0 +1,301 @@
+import { createApi, fetchBaseQuery } from "@reduxjs/toolkit/query/react";
+import { RootState } from "../../app/store";
+import { DOCKER_CONTAINER_ACTION, DOCKER_CONTAINER_LIST } from "./consts";
+
```
Comment (MarcMcIntosh) [inline src/services/refact/docker.ts:4]: There might be some cache issues here,

File: src/features/Integrations/renderIntegrationFormField.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+  values,
+  fieldKey,
+  f_type,
+}: {
+  fieldKey: string;
+  values: Integration["integr_values"];
+  field: IntegrationField<NonNullable<IntegrationPrimitive>>;
+  f_type: "bool" | "int" | "string";
+}) => {
+  if (values[fieldKey]) {
```
Comment (MarcMcIntosh) [inline src/features/Integrations/renderIntegrationFormField.tsx:33]: Could be a falsy value `0, null ""` ?
```
if(fieldKey in values)  {

}
```
or 
```
if(values[fieldKey] !== undefined) {}
```

**Review by MarcMcIntosh — COMMENTED** [review]
Left some comments.

File: src/components/IntegrationsView/JSONFormStyles.css (hunk 2 above)
Comment (alashchev17) [reply src/components/IntegrationsView/JSONFormStyles.css:1]: Those ones should be removed, it's a legacy from the json schema parsing library

File: src/features/Integrations/renderIntegrationFormField.tsx (hunk 6 above)
Comment (alashchev17) [reply src/features/Integrations/renderIntegrationFormField.tsx:33]: Good point 👍🏻 
Done, pushed

File: src/services/refact/docker.ts (hunk 4 above)
Comment (alashchev17) [reply src/services/refact/docker.ts:294]: To be honest, we don't even render them, the only requirement for `labels` and `ports` is to be an object (`NonNullable<unknown>`)

File: src/features/Chat/Thread/reducer.ts (hunk 3 above)
Comment (alashchev17) [reply src/features/Chat/Thread/reducer.ts:200]: `TODO: find out about tool use` this one also?

File: src/components/IntegrationsView/IntegrationsView.tsx (hunk 1 above)
Comment (alashchev17) [reply src/components/IntegrationsView/IntegrationsView.tsx:82]: It was your comment 😁

File: src/services/refact/docker.ts (hunk 5 above)
Comment (alashchev17) [reply src/services/refact/docker.ts:4]: Added TODO to test it and adjust in the future

File: src/features/Chat/Thread/reducer.ts (hunk 3 above)
Comment (MarcMcIntosh) [reply src/features/Chat/Thread/reducer.ts:200]: no, I think that's still up in the air with the new chat_mode object in the meta property for chat requests.

**Review by MarcMcIntosh — APPROVED** [review]

File: src/components/IntegrationsView/IntegrationCard.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+export const IntegrationCard: FC<IntegrationCardProps> = ({
+  integration,
+  handleIntegrationShowUp,
+  isInline = false,
+}) => {
+  const integrationLogo = INTEGRATIONS_WITH_TERMINAL_ICON.includes(
+    integration.integr_name.split("_")[0],
+  )
+    ? `/integrations/cmdline.png`
```
[the remark this answers is not in the document]
Comment (alashchev17) [reply src/components/IntegrationsView/IntegrationCard.tsx:29]: will do right now

**Pull request merged**